File: backend/utils/permissions.py
"""
OutreachRoute Pro — Permission Helpers

Functions to check whether a user has permission to access a given record.
These are the backend enforcement layer — frontend also enforces via permissions.js.
"""

from models.user import User


ADMIN_ROLES = {"master_admin", "national_admin", "regional_admin", "state_admin", "local_admin"}


def is_admin(user: User) -> bool:
    """Return True if the user has any admin-level role."""
    return user.role in ADMIN_ROLES


def can_view_applicant(user: User, applicant) -> bool:
    """
    Return True if the user is allowed to view the given applicant.
    - Admins can see all applicants in their geographic scope.
    - OA users can only see applicants assigned to them.
    """
    if is_admin(user):
        return True
    return applicant.assigned_oa_id == user.id


def can_edit_applicant(user: User, applicant) -> bool:
    """Return True if the user can edit this applicant record."""
    if is_admin(user):
        return True
    return applicant.assigned_oa_id == user.id


def can_manage_users(user: User) -> bool:
    """Return True if the user can create/edit/deactivate other users."""
    return is_admin(user)


def can_view_all_territory(user: User) -> bool:
    """Return True if the user has master or national admin access."""
    return user.role in {"master_admin", "national_admin"}

File: frontend/src/utils/permissions.js
/**
 * OutreachRoute Pro — Frontend Permission Utilities
 *
 * These mirror the backend permission rules so the UI can hide/show
 * elements based on the current user's role.
 * The backend always re-validates — these are UI helpers only.
 */

const ADMIN_ROLES = new Set([
  "master_admin",
  "national_admin",
  "regional_admin",
  "state_admin",
  "local_admin",
]);

/**
 * Return true if the user has any admin-level role.
 */
export function isAdmin(user) {
  return user ? ADMIN_ROLES.has(user.role) : false;
}

/**
 * Return true if the user is a master admin.
 */
export function isMasterAdmin(user) {
  return user?.role === "master_admin";
}

/**
 * Return true if the user can manage other users (create/edit/deactivate).
 */
export function canManageUsers(user) {
  return isAdmin(user);
}

/**
 * Return true if the user can view the team performance dashboard.
 */
export function canViewTeamPerformance(user) {
  return isAdmin(user);
}

/**
 * Return true if the user can manage territory assignments.
 */
export function canManageTerritories(user) {
  return isAdmin(user);
}

/**
 * Return true if the user can view a given applicant.
 * Admins see all; OA users only see assigned applicants.
 */
export function canViewApplicant(user, applicant) {
  if (!user || !applicant) return false;
  if (isAdmin(user)) return true;
  return applicant.assigned_oa_id === user.id;
}

/**
 * Return true if the user can edit a given applicant's record.
 */
export function canEditApplicant(user, applicant) {
  return canViewApplicant(user, applicant);
}
